Replace pkgConfig calls to configure this project with CMake calls

fbudin69500 opened the pull request:
This project now uses `find_package()` commands to find its dependencies.
It also exports its targets both with  *.pc files and *.cmake files, which
means that it can be consumed both with standard CMake `find_package()`
commands or with `PkgConfig` commands.

jamiesnape commented on CMakeLists.txt:
```diff
@@ -1,4 +1,4 @@
-cmake_minimum_required(VERSION 2.8.5)
+cmake_minimum_required(VERSION 3.6)
```
We need this to work for 3.5.1 if at all possible.

jamiesnape commented on CMakeLists.txt:
```diff
@@ -65,26 +59,26 @@ set(default_cmake_args
-  CMAKE_CACHE_ARGS ${default_cmake_args})
-checkbuild(bot2-core)
+  CMAKE_CACHE_ARGS ${default_cmake_args}
+  BUILD_ALWAYS)
```
`BUILD_ALWAYS 1`

jamiesnape commented on bot2-core/cmake/lcmtypes.cmake:
```diff
@@ -279,8 +273,8 @@ function(lcmtypes_build_cpp)
-    include_directories(BEFORE ${PROJECT_SOURCE_DIR}/lcmtypes/cpp) #TODO: don't think this is necessary
-    include_directories(${LCM_INCLUDE_DIRS})
+    #include_directories(BEFORE ${PROJECT_SOURCE_DIR}/lcmtypes/cpp) #TODO: don't think this is necessary
+    #include_directories(${LCM_INCLUDE_DIRS})
```
Delete.

jamiesnape commented on bot2-core/cmake/lcmtypes.cmake:
```diff
@@ -170,19 +158,22 @@ function(lcmtypes_build_c)
 
-    include_directories(BEFORE ${PROJECT_SOURCE_DIR}/lcmtypes/c) #TODO: don't think this is necessary
-    include_directories(${LCM_INCLUDE_DIRS})
+#    include_directories(${LCM_INCLUDE_DIRS})
```
Delete.

jamiesnape commented on bot2-core/cmake/pods.cmake:
```diff
@@ -339,22 +365,22 @@ macro(pods_config_search_paths)
-        include_directories(BEFORE ${INCLUDE_OUTPUT_PATH})
-        include_directories(${INCLUDE_INSTALL_PATH})
+#        include_directories(BEFORE ${INCLUDE_OUTPUT_PATH})
+#        include_directories(${INCLUDE_INSTALL_PATH})
```
Delete.

jamiesnape commented on bot2-frames/cmake/lcmtypes.cmake:
```diff
@@ -170,19 +158,22 @@ function(lcmtypes_build_c)
 
-    include_directories(BEFORE ${PROJECT_SOURCE_DIR}/lcmtypes/c) #TODO: don't think this is necessary
-    include_directories(${LCM_INCLUDE_DIRS})
+#    include_directories(${LCM_INCLUDE_DIRS})
```
Delete.

jamiesnape commented on bot2-frames/cmake/lcmtypes.cmake:
```diff
@@ -170,19 +158,22 @@ function(lcmtypes_build_c)
     set_target_properties(${libname} PROPERTIES CLEAN_DIRECT_OUTPUT 1)
+    target_link_libraries(${libname} ${lcm_namespace}lcm-coretypes ${lcm_namespace}lcm)
+    target_include_directories(${libname} PUBLIC
+#        $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/lcmtypes/c>
```
Delete.

jamiesnape commented on bot2-frames/cmake/lcmtypes.cmake:
```diff
@@ -279,8 +273,8 @@ function(lcmtypes_build_cpp)
-    include_directories(BEFORE ${PROJECT_SOURCE_DIR}/lcmtypes/cpp) #TODO: don't think this is necessary
-    include_directories(${LCM_INCLUDE_DIRS})
+    #include_directories(BEFORE ${PROJECT_SOURCE_DIR}/lcmtypes/cpp) #TODO: don't think this is necessary
+    #include_directories(${LCM_INCLUDE_DIRS})
```
Delete.

jamiesnape commented on bot2-frames/cmake/pods.cmake:
```diff
@@ -306,7 +308,7 @@ macro(pods_use_pkg_config_packages target)
     string(STRIP ${_pods_pkg_ldflags} _pods_pkg_ldflags)
     #    message("ldflags: ${_pods_pkg_ldflags}")
-    include_directories(${_pods_pkg_include_flags})
+#    include_directories(${_pods_pkg_include_flags})
```
Delete.

jamiesnape commented on bot2-frames/cmake/pods.cmake:
```diff
@@ -339,22 +365,22 @@ macro(pods_config_search_paths)
-        include_directories(BEFORE ${INCLUDE_OUTPUT_PATH})
-        include_directories(${INCLUDE_INSTALL_PATH})
+#        include_directories(BEFORE ${INCLUDE_OUTPUT_PATH})
+#        include_directories(${INCLUDE_INSTALL_PATH})
```
Delete.

jamiesnape commented on bot2-lcm-utils/cmake/pods.cmake:
```diff
@@ -306,7 +308,7 @@ macro(pods_use_pkg_config_packages target)
     string(STRIP ${_pods_pkg_ldflags} _pods_pkg_ldflags)
     #    message("ldflags: ${_pods_pkg_ldflags}")
-    include_directories(${_pods_pkg_include_flags})
+#    include_directories(${_pods_pkg_include_flags})
```
Delete.

jamiesnape commented on bot2-lcm-utils/cmake/pods.cmake:
```diff
@@ -339,22 +365,22 @@ macro(pods_config_search_paths)
-        include_directories(BEFORE ${INCLUDE_OUTPUT_PATH})
-        include_directories(${INCLUDE_INSTALL_PATH})
+#        include_directories(BEFORE ${INCLUDE_OUTPUT_PATH})
+#        include_directories(${INCLUDE_INSTALL_PATH})
```
Delete.

jamiesnape commented on bot2-lcmgl/cmake/lcmtypes.cmake:
```diff
@@ -170,19 +158,22 @@ function(lcmtypes_build_c)
 
-    include_directories(BEFORE ${PROJECT_SOURCE_DIR}/lcmtypes/c) #TODO: don't think this is necessary
-    include_directories(${LCM_INCLUDE_DIRS})
+#    include_directories(${LCM_INCLUDE_DIRS})
```
Delete.

jamiesnape commented on bot2-lcmgl/cmake/lcmtypes.cmake:
```diff
@@ -279,8 +273,8 @@ function(lcmtypes_build_cpp)
-    include_directories(BEFORE ${PROJECT_SOURCE_DIR}/lcmtypes/cpp) #TODO: don't think this is necessary
-    include_directories(${LCM_INCLUDE_DIRS})
+    #include_directories(BEFORE ${PROJECT_SOURCE_DIR}/lcmtypes/cpp) #TODO: don't think this is necessary
+    #include_directories(${LCM_INCLUDE_DIRS})
```
Delete.

jamiesnape commented on bot2-lcmgl/cmake/pods.cmake:
```diff
@@ -306,7 +308,7 @@ macro(pods_use_pkg_config_packages target)
     string(STRIP ${_pods_pkg_ldflags} _pods_pkg_ldflags)
     #    message("ldflags: ${_pods_pkg_ldflags}")
-    include_directories(${_pods_pkg_include_flags})
+#    include_directories(${_pods_pkg_include_flags})
```
Delete.

jamiesnape commented on bot2-param/cmake/pods.cmake:
```diff
@@ -306,7 +308,7 @@ macro(pods_use_pkg_config_packages target)
     string(STRIP ${_pods_pkg_ldflags} _pods_pkg_ldflags)
     #    message("ldflags: ${_pods_pkg_ldflags}")
-    include_directories(${_pods_pkg_include_flags})
+#    include_directories(${_pods_pkg_include_flags})
```
Delete.

jamiesnape commented on bot2-param/cmake/pods.cmake:
```diff
@@ -339,22 +365,22 @@ macro(pods_config_search_paths)
-        include_directories(BEFORE ${INCLUDE_OUTPUT_PATH})
-        include_directories(${INCLUDE_INSTALL_PATH})
+#        include_directories(BEFORE ${INCLUDE_OUTPUT_PATH})
+#        include_directories(${INCLUDE_INSTALL_PATH})
```
Delete.

jamiesnape commented on bot2-procman/cmake/lcmtypes.cmake:
```diff
@@ -170,19 +158,22 @@ function(lcmtypes_build_c)
 
-    include_directories(BEFORE ${PROJECT_SOURCE_DIR}/lcmtypes/c) #TODO: don't think this is necessary
-    include_directories(${LCM_INCLUDE_DIRS})
+#    include_directories(${LCM_INCLUDE_DIRS})
```
Delete.

jamiesnape commented on bot2-procman/cmake/lcmtypes.cmake:
```diff
@@ -279,8 +273,8 @@ function(lcmtypes_build_cpp)
-    include_directories(BEFORE ${PROJECT_SOURCE_DIR}/lcmtypes/cpp) #TODO: don't think this is necessary
-    include_directories(${LCM_INCLUDE_DIRS})
+    #include_directories(BEFORE ${PROJECT_SOURCE_DIR}/lcmtypes/cpp) #TODO: don't think this is necessary
+    #include_directories(${LCM_INCLUDE_DIRS})
```
Delete.

jamiesnape commented on bot2-procman/cmake/pods.cmake:
```diff
@@ -306,7 +308,7 @@ macro(pods_use_pkg_config_packages target)
     string(STRIP ${_pods_pkg_ldflags} _pods_pkg_ldflags)
     #    message("ldflags: ${_pods_pkg_ldflags}")
-    include_directories(${_pods_pkg_include_flags})
+#    include_directories(${_pods_pkg_include_flags})
```
Delete.

jamiesnape commented on bot2-procman/cmake/pods.cmake:
```diff
@@ -339,22 +365,22 @@ macro(pods_config_search_paths)
-        include_directories(BEFORE ${INCLUDE_OUTPUT_PATH})
-        include_directories(${INCLUDE_INSTALL_PATH})
+#        include_directories(BEFORE ${INCLUDE_OUTPUT_PATH})
+#        include_directories(${INCLUDE_INSTALL_PATH})
```
Delete.

jamiesnape commented on bot2-vis/cmake/pods.cmake:
```diff
@@ -306,7 +308,7 @@ macro(pods_use_pkg_config_packages target)
     string(STRIP ${_pods_pkg_ldflags} _pods_pkg_ldflags)
     #    message("ldflags: ${_pods_pkg_ldflags}")
-    include_directories(${_pods_pkg_include_flags})
+#    include_directories(${_pods_pkg_include_flags})
```
Delete.

jamiesnape commented on bot2-core/src/bot_core/CMakeLists.txt:
```diff
@@ -1,24 +1,27 @@
+target_include_directories(${POD_NAME} PUBLIC
+  $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/src>
+  $<INSTALL_INTERFACE:include> )
+target_compile_options(${POD_NAME} PUBLIC -std=gnu99)
```
Set `C_STANDARD` target property instead?

jamiesnape commented on bot2-frames/src/CMakeLists.txt:
```diff
@@ -1,26 +1,38 @@
+    $<BUILD_INTERFACE:${PROJECT_BINARY_DIR}/include>
+    $<INSTALL_INTERFACE:include>
+)
+target_compile_options(${POD_NAME} PUBLIC -std=gnu99)
```
Set `C_STANDARD` target property instead?

jamiesnape commented on bot2-frames/src/renderer/CMakeLists.txt:
```diff
@@ -1,16 +1,20 @@
-    ${OPENGL_LIBRARIES})
-list(APPEND REQUIRED_PACKAGES bot2-frames)
-
+target_compile_options(bot2-frames-renderers PUBLIC -std=gnu99)
```
Set `C_STANDARD` target property instead?

jamiesnape commented on bot2-vis/src/wavefront-viewer/CMakeLists.txt:
```diff
@@ -1,9 +1,7 @@
     main.c
     renderer_wavefront.c)
+target_link_libraries(bot-wavefront-viewer bot2-vis ${GTK2_LIBRARIES} OpenGL::GL ${GLUT_LIBRARIES})
+target_compile_options(bot-wavefront-viewer PRIVATE -std=gnu99)
```
Set `C_STANDARD` target property instead?

jamiesnape commented on bot2-vis/src/testers/CMakeLists.txt:
```diff
@@ -1,11 +1,10 @@
     add_executable(${tester} "${tester}.c")
-    pods_use_pkg_config_packages(${tester} bot2-vis)  # need headers as well as bot2-vis target
+    target_link_libraries(${tester} bot2-vis GLib2::glib ${GTK2_LIBRARIES} OpenGL::GL OpenGL::GLU)
+    target_compile_options(${tester} PRIVATE -std=gnu99)
```
Set `C_STANDARD` target property instead?

jamiesnape commented on bot2-vis/src/rwx-viewer/CMakeLists.txt:
```diff
@@ -1,9 +1,7 @@
     main.c
     renderer_rwx.c)
+target_link_libraries(bot-rwx-viewer bot2-vis ${GTK2_LIBRARIES} OpenGL::GL ${GLUT_LIBRARIES})
+target_compile_options(bot-rwx-viewer PRIVATE -std=gnu99)
```
Set `C_STANDARD` target property instead?

jamiesnape commented on bot2-vis/src/bot_vis/CMakeLists.txt:
```diff
@@ -4,31 +4,54 @@ file(GLOB h_files *.h)
+    ${PNG_INCLUDE_DIRS}
+    ${JPEG_INCLUDE_DIR}
+)
+target_compile_options(bot2-vis PUBLIC -std=gnu99 PRIVATE ${PNG_DEFINITIONS})
```
Set `C_STANDARD` target property instead?

jamiesnape commented on bot2-procman/src/deputy/CMakeLists.txt:
```diff
@@ -1,16 +1,14 @@
     procman.c
     procman_deputy.c
     signal_pipe.c)
+target_compile_options(bot-procman-deputy PRIVATE -std=gnu99)
```
Set `C_STANDARD` target property instead?

jamiesnape commented on bot2-param/src/param_tester/CMakeLists.txt:
```diff
@@ -1,14 +1,12 @@
-pods_use_pkg_config_packages(bot-param-dump lcm glib-2.0)
-target_link_libraries(bot-param-dump bot2-param-client)
+target_link_libraries(bot-param-dump bot2-param-client ${lcm_namespace}lcm GLib2::glib lcmtypes_bot2-param)
+target_compile_options(bot-param-dump PRIVATE -std=gnu99)
```
Set `C_STANDARD` target property instead?

jamiesnape commented on bot2-param/src/param_server/CMakeLists.txt:
```diff
@@ -1,15 +1,13 @@
-pods_use_pkg_config_packages(bot-param-server lcm glib-2.0)
-target_link_libraries(bot-param-server bot2-param-client)
+target_link_libraries(bot-param-server bot2-param-client ${lcm_namespace}lcm GLib2::glib lcmtypes_bot2-param)
+target_compile_options(bot-param-server PRIVATE -std=gnu99)
```
Set `C_STANDARD` target property instead?

jamiesnape commented on bot2-param/src/param_client/CMakeLists.txt:
```diff
@@ -1,12 +1,12 @@
+target_include_directories(bot2-param-client PUBLIC
+  $<BUILD_INTERFACE:${PROJECT_BINARY_DIR}/include>
+  $<INSTALL_INTERFACE:include> )
+target_compile_options(bot2-param-client PUBLIC -std=gnu99)
```
Set `C_STANDARD` target property instead?

jamiesnape commented on bot2-lcmgl/src/lcmgl-viewer/CMakeLists.txt:
```diff
@@ -1,13 +1,13 @@
-pods_use_pkg_config_packages(bot-lcmgl-viewer bot2-vis glib-2.0 gthread-2.0)
+target_link_libraries(bot-lcmgl-viewer bot2-lcmgl-renderer libbot::bot2-vis libbot::bot2-core ${GTK2_LIBRARIES})
+target_include_directories(bot-lcmgl-viewer PUBLIC ${GTK2_INCLUDE_DIRS})
+target_compile_options(bot-lcmgl-viewer PRIVATE -std=gnu99)
```
Set `C_STANDARD` target property instead?

jamiesnape commented on bot2-lcmgl/src/bot_lcmgl_client/CMakeLists.txt:
```diff
@@ -1,29 +1,23 @@
-list(APPEND REQUIRED_PACKAGES lcmtypes_bot2-lcmgl)        
+target_link_libraries(${libname} PUBLIC ${lcm_namespace}lcm PRIVATE ${ZLIB_LIBRARIES} lcmtypes_bot2-lcmgl)
+target_include_directories(${libname} PRIVATE ${ZLIB_INCLUDE_DIRS})
+target_compile_options(${libname} PUBLIC -std=gnu99)
```
Set `C_STANDARD` target property instead?

jamiesnape commented on bot2-lcmgl/cmake/pods.cmake:
```diff
@@ -339,22 +365,22 @@ macro(pods_config_search_paths)
-        include_directories(BEFORE ${INCLUDE_OUTPUT_PATH})
-        include_directories(${INCLUDE_INSTALL_PATH})
+#        include_directories(BEFORE ${INCLUDE_OUTPUT_PATH})
+#        include_directories(${INCLUDE_INSTALL_PATH})
```
Delete.

jamiesnape commented on bot2-lcm-utils/src/who/CMakeLists.txt:
```diff
@@ -1,10 +1,7 @@
 add_executable(bot-lcm-who
     lcm-who.c
     signal_pipe.c)
+target_compile_options(bot-lcm-who PRIVATE -std=gnu99)
```
Set `C_STANDARD` target property instead?

jamiesnape commented on bot2-lcm-utils/src/logsplice/CMakeLists.txt:
```diff
@@ -1,10 +1,7 @@
-
 add_executable(bot-lcm-logsplice
     lcm-logsplice.c)
+target_compile_options(bot-lcm-logsplice PRIVATE -std=gnu99)
```
Set `C_STANDARD` target property instead?

jamiesnape commented on bot2-lcm-utils/src/logfilter/CMakeLists.txt:
```diff
@@ -1,9 +1,6 @@
-
 add_executable(bot-lcm-logfilter
     lcm-logfilter.c)
+target_compile_options(bot-lcm-logfilter PRIVATE -std=gnu99)
```
Set `C_STANDARD` target property instead?

jamiesnape commented on bot2-frames/src/renderer/CMakeLists.txt:
```diff
@@ -20,21 +24,23 @@ set_target_properties(bot2-frames-renderers PROPERTIES SOVERSION 1)
-                                         glib-2.0 gthread-2.0)
+target_link_libraries(test-viewer ${GLUT_LIBRARIES} bot2-frames-renderers libbot::bot2-vis libbot::bot2-lcmgl-renderer)
+target_include_directories(test-viewer PRIVATE ${GLUT_INCLUDE_DIR} ${GTK2_INCLUDE_DIRS})
+target_compile_options(test-viewer PRIVATE -std=gnu99)
```
Set `C_STANDARD` target property instead?

jamiesnape commented on bot2-frames/src/test/CMakeLists.txt:
```diff
@@ -1,10 +1,7 @@
-target_link_libraries(coord-frames-test bot2-frames)
-
+target_link_libraries(coord-frames-test ${POD_NAME} libbot::bot2-param-client)
+target_compile_options(coord-frames-test PRIVATE -std=gnu99)
```
Set `C_STANDARD` target property instead?

jamiesnape commented on CMakeLists.txt:
```diff
@@ -65,26 +59,26 @@ set(default_cmake_args
-  CMAKE_CACHE_ARGS ${default_cmake_args})
-checkbuild(bot2-procman)
+  CMAKE_CACHE_ARGS ${default_cmake_args}
+  BUILD_ALWAYS)
```
`BUILD_ALWAYS 1`

jamiesnape commented on CMakeLists.txt:
```diff
@@ -117,12 +111,18 @@ if (WITH_BOT_VIS)
-    CMAKE_CACHE_ARGS ${default_cmake_args} -DUSE_BOT_VIS:BOOL=${WITH_BOT_VIS})
-  checkbuild(bot2-lcmgl)
+    CMAKE_CACHE_ARGS ${default_cmake_args} -DUSE_BOT_VIS:BOOL=${WITH_BOT_VIS}
+    BUILD_ALWAYS)
```
`BUILD_ALWAYS 1`

jamiesnape commented:
Can you install FindOpenGL to `lib/cmake/${POD_NAME}/modules`? Matches the convention we are going to use in https://github.com/RobotLocomotion/drake/pull/7323.

Also change `"${var}" STREQUAL "foo"` to `var STREQUAL "foo"` and change `if (` to `if(`. Keeps __@&#x202f;mwoehlke-kitware__ happy.

fbudin69500 commented:
@jamiesnape Actually, because `"${var}" STREQUAL "foo"` is used in a macro and `var` is a macro argument, we need to write it like this. See [macro documentation](https://cmake.org/cmake/help/v3.10/command/macro.html)

jamiesnape commented:
It is a macro? I did not notice. That is fine then.

fbudin69500 commented:
@jamiesnape Anything else to modify before this is merged? It needs your approval.

jamiesnape commented:
Did you fix the installed location of `FindOpenGL`?

jamiesnape commented on bot2-core/cmake/lcmtypes.cmake:
```diff
@@ -466,20 +449,29 @@ function(lcmtypes_install_types)
-    	return()
+    find_package(lcm REQUIRED)
+    if(TARGET lcm::lcm-gen)
+      set(lcm_namespace "lcm::")
```
If a recent version of LCM is being used, it will set `LCM_NAMESPACE`. Use that everywhere you are using `lcm::` and by default `LCM_NAMESPACE` is set to `""` so the TARGET is called simply `lcm-gen`.

jamiesnape commented on bot2-core/cmake/lcmtypes.cmake:
```diff
@@ -466,20 +449,32 @@ function(lcmtypes_install_types)
+      message(FATAL_ERROR "lcm-gen not found")
     endif()
-    
+    if(TARGET lcm::lcm-java)
```
This block can be simplified with `LCM_NAMESPACE` as well, I think.

Commit:
Replace pkgConfig calls to configure this project with CMake calls

This project now uses `find_package()` commands to find its dependencies.
It also exports its targets both with  *.pc files and *.cmake files, which
means that it can be consumed both with standard CMake `find_package()`
commands or with `PkgConfig` commands.
